### DISCUSSION TITLE
Upgrade to java 11/17 ?

### DISCUSSION THREAD

**Discussion by gitmotte** [JSQLParser/JSqlParser#1408] [General]
upgrade to one of the newer lts versions (11/17) to benefit from new language features?

https://www.oracle.com/java/technologies/java-se-support-roadmap.html

**Comment by manticore-projects** [comment]
Greetings!

It was a great and reasonable idea in general, but unfortunately the reality is never great or reasonable, but cruel and dirty: Our customers still run CentOs 7 boxes with JDK8 only and are reluctant to touch anything because they would need to upgrade OS and Oracle DB also, while they achieve nothing.

The boxes run without an internet connection and very stable. Uptime more than 1 year. How could we force them to migrate to another JDK now? 

Lets discuss that again in 5 years maybe.

**Reply by gitmotte** [reply]
Not at all. The boxes can continue to run without downtime. There are still old releases for java8. Apparently there is no need for further development?

**Reply by manticore-projects** [reply]
As long as we achieve compatibility with the Java 8 class format/version, I would not object.
However, what features of Java11/17 would really matter? Only MultiLine Strings would come to my mind, maybe the pattern based `instance of`. I do not see a major advantage yet?

**Reply by gitmotte** [reply]
there are some interesting new features.

https://advancedweb.hu/a-categorized-list-of-all-java-and-jvm-features-since-jdk-8-to-17/

i like many of them, some examples up to jdk11
 * Introduction of **var** to make local variable declarations less ceremonious (jdk11)
 * module - system to hide internal apis (Project Jigsaw) (jdk9)
 * String enhancements, like isBlank, lines, repeat and strip (jdk11)
 * Convenience Factory Methods for Collections to ease the pain of not having collection literals (jdk9)
 * ...

**Reply by manticore-projects** [reply]
Thank you for the feedback.

> i like many of them, some examples up to jdk11

You might be a much better programmer than I am, but I still find it interesting how opinion can differ:

>     * Introduction of **var** to make local variable declarations less ceremonious (jdk11)
I fail to see much benefit -- except occupying another keyword.

>     * module - system to hide internal apis (Project Jigsaw) (jdk9)
I hate this most! So much hassle and breaking stuff for nothing.

>     * String enhancements, like isBlank, lines, repeat and strip (jdk11)
Certainly useful, but if you are truly interested in String manipulation then better engage Groovy (which has had Multi Line Strings and proper String Templates for 15 years or so).

>     * Convenience Factory Methods for Collections to ease the pain of not having collection literals (jdk9)
Again I fail to see much benefit.

I know it is just an uneducated opinion of an amateur, but to me the only interesting (new) features are GraalVM with Python/R integration (which of course do not work for the Forecast packages :-)) and Native Image and Multiline Strings and Math.FMA().

That said, its just me. I definitely need Java 8 Class compatibility and will need to run JSQLParser on ancient Java 8 JVMs, as long as this works out I do not bother.

**Reply by manticore-projects** [reply]
And sorry for derailing!

**Reply by gitmotte** [reply]
the derailment is directing itself ;)

diversity of opinion is a given when there are several cooks/guests. i don't think much of making a customer believe in a perfect world - sometimes customers have to be given the choice that there will only be more necessary bugfixes if they don't change over. 

an upgrade with big version jumps will be all the more expensive than many small upgrades.

**Reply by manticore-projects** [reply]
> the derailment is directing itself ;)
> 
> diversity of opinion is a given when there are several cooks/guests. i don't think much of making a customer believe in a perfect world - sometimes customers have to be given the choice that there will only be more necessary bugfixes if they don't change over.
> 
> an upgrade with big version jumps will be all the more expensive than many small upgrades.

So what are all those Mainframes and  Cobol programs good for?
Can I switch these off?

**Reply by gitmotte** [reply]
^^ there are also different cobol versions. apart from that, cobol programs are now ported to java for expensive money or expensive cobol developers are brought out of retirement ...

**Reply by manticore-projects** [reply]
On a more serious note: All Oracle DBs we see running here are Oracle 11.2 mostly on Centos 7 or 6.8 or AIX. The JVM is our smallest problem by far.

They just took that stuff and move it into some Microsoft Clouds, where it will rot forever -- as long as the P&L looks good. The hellish version of an eternal software heaven.

**Reply by manticore-projects** [reply]
> cobol programs are now ported to java

And you are sure, that would be Java17 Bytecode? :-D
(Sorry for trolling.)

**Reply by gitmotte** [reply]
trolling must be your main occupation, or is the desperation already so great?
i can only speak of my customers, who gratefully pay for every upgrade to keep the software up to date and maintainable. and i'm not talking about always using the latest, but the customers have already fled java8 for support and license reasons alone ;)
